### PR TITLE
Major aggregator documentation refactor and minor aggregator refactor

### PR DESCRIPTION
**Pull request by ashelkovnykov**
- Updated all aggregator documentation
- Updated BroadcastWrapper documentation
- Removed Coefficients type from ObjectiveFunction
- Moved coefficient and vector broadcast logic into aggregators
- Added checks on input to NormalizationContext

### CODE REVIEW

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/glm/HessianVectorAggregator.scala [hunk 1]
```diff
@@ -15,159 +15,276 @@
...
-   * @param coef The current model coefficients
-   * @param multiplyVector The Hessian multiplication vector
-   * @param normalizationContext The normalization context
+   * H_jk = ∂^2^ L(w, X, y, u, o) / (∂ w_j)(∂ w_k)
+   *      = \sum_i ∂^2^ (u_i * l(z_i, y_i)) / (∂ w_j)(∂ w_k)
+   *      = \sum_i (∂^2^ l(z_i, y_i) / (∂ z_i)^2^) * u_i * X_i_j * X_i_k
+   *
+   * Let v = H * d, where d is some vector. Then v_j = H_j * d, where:
+   *
+   * H_j = /sum_i H_i_j         =>
```
Comment (yunboouyang) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/glm/HessianVectorAggregator.scala]: Should the summation notation be unified as `\sum`?

Comment (ashelkovnykov) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/function/glm/HessianVectorAggregator.scala]: That's a typo - will fix

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala [hunk 2]
```diff
@@ -185,36 +182,48 @@ trait L2RegularizationTwiceDiff extends TwiceDiffFunction with L2RegularizationD
...
   /**
-   * Compute the Hessian matrix for the function with L2 regularization over the given data for the given model
-   * coefficients.
-   *
-   * @param input The given data over which to compute the diagonal of the Hessian matrix
-   * @param coefficients The model coefficients used to compute the diagonal of the Hessian matrix
-   * @return The computed Hessian matrix
+   * Compute H * d (where H is the Hessian matrix and d is some vector) over the given data for the given model
+   * coefficients. This is a special helper function which computes H * d more efficiently than computing the entire
+   * Hessian matrix and then multiplying it with d.
```
Comment (yunboouyang) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala]: Good documentation!

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala [hunk 3]
```diff
@@ -185,36 +182,48 @@ trait L2RegularizationTwiceDiff extends TwiceDiffFunction with L2RegularizationD
...
+  protected def l2RegHessianMatrix(dimension: Int): DenseMatrix[Double] =
+    l2RegWeight * DenseMatrix.eye[Double](dimension)
 
   /**
    * Compute the Hessian diagonal of the L2 regularization term.
    *
    * @return The Hessian diagonal of the L2 regularization term
    */
-  protected def l2RegHessianDiagonal: Double = l2RegWeight
+  protected def l2RegHessianDiagonal(dimension: Int): Vector[Double] = DenseVector.fill(dimension, l2RegWeight)
```
Comment (cmjiang) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala]: Miss the param `dimension` in the comment.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/glm/PointwiseLossFunction.scala [hunk 4]
```diff
@@ -15,31 +15,34 @@
...
- *  - z is the margin
  *  - y is the label
  *
- * It is the negative of the log-likelihood for one data point, except for an irrelevant constant term.
+ * l(z, y) is the negative of the log-likelihood (discounting the irrelevant constant term).
+ *
+ * For example:
  *
- * For example, for linear regression, l(z, y) = 1/2 (z - y)^2^.
+ *    l(z, y) = 1/2 (z - y)^2^.
```
Comment (cmjiang) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/glm/PointwiseLossFunction.scala]: Extra period at the end of this line.

**Pull request merged**